### PR TITLE
Represent strings as re-tagged vectors of chars

### PR DESCRIPTION
**Pull request by eholk**
This is a redo of #55 building on the vector support in #65.

This should have addressed most of the comments in #55, but it wouldn't hurt to double check.

Unfortunately, this seems to hurt compiler performance quite a bit. I'm seeing compile times in the high 5 seconds now, where they used to be in the low 3 second range.

### CODE REVIEW

File: schism/compiler.ss [hunk 1]
```diff
@@ -222,29 +256,55 @@
          (or (char-numeric? c) (char-between c #\a #\f)))
        (define (char-ci<? c1 c2)
          (< (char->integer c1) (char->integer c2)))
+       (define (make-string n)
+         (%set-tag (make-vector n #\0) ,(string-tag)))
+       (define (string-length s)
+         (if (string? s)
+             (vector-length (%set-tag s ,(vector-tag)))
+             (error 'string-ref "Invalid arguments")))
```
Comment (wingo) [inline schism/compiler.ss]: 'string-length

File: schism/compiler.ss [hunk 2]
```diff
@@ -222,29 +256,55 @@
...
+         (if (and (string? s)
+                  (< i (string-length s)))
+             (vector-ref (%set-tag s ,(vector-tag)) i)
+             (error 'string-ref "Invalid arguments")))
+       (define (string-set! s i c)
+         (if (and (string? s)
+                  (< i (string-length s))
+                  (char? c))
+             (vector-set! (%set-tag s ,(vector-tag)) i c)
+             (error 'string-ref "Invalid arguments")))
```
Comment (wingo) [inline schism/compiler.ss]: 'string-set!

File: schism/compiler.ss [hunk 3]
```diff
@@ -953,6 +1015,98 @@
...
+  ;; Lower strings          ;;
+  ;; ====================== ;;
+
+  ;; lower-strings removes all string literals. At the moment, it does
+  ;; this by replacing string literals with code to allocate and build
+  ;; a string, but eventually we will be able to do this ahead of time
+  ;; and store the strings in a static data section.
+
+  (define (lower-strings fn*)
+    (map (lambda (fn) (lower-strings-fn fn)) fn*))
```
Comment (wingo) [inline schism/compiler.ss]: I think you can just map lower-strings-fn directly, without the lambda

File: schism/compiler.ss [hunk 4]
```diff
@@ -953,6 +1015,98 @@
...
+          (cond
+           ((eq? tag 'null) expr)
+           ((eq? tag 'bool) expr)
+           ((eq? tag 'char) expr)
+           ((eq? tag 'begin)
+	    (cons 'begin (map lower-strings-expr (cdr expr))))
+           ((eq? tag 'var) expr)
+           ((eq? tag 'number) expr)
+           ((eq? tag 'eof-object) expr)
+	   ((eq? tag 'string) (lower-string (cadr expr)))
```
Comment (wingo) [inline schism/compiler.ss]: fix indentation

File: schism/compiler.ss [hunk 5]
```diff
@@ -953,6 +1015,98 @@
...
+           ((eq? tag 'call)
+            (cons 'call (cons (cadr expr)
+			      (map lower-strings-expr (cddr expr)))))
```
Comment (wingo) [inline schism/compiler.ss]: just a question: do you mean to have tabs here?

File: schism/compiler.ss [hunk 6]
```diff
@@ -953,6 +1015,98 @@
...
+               (error 'lower-strings-expr "malformed expr"))))
+  (define (lower-strings-bindings bindings)
+    (map (lambda (b)
+	   (let ((name (car b))
+		 (expr (cadr b)))
+	     `(,name ,(lower-strings-expr expr))))
+	 bindings))
+  (define (lower-string s)
+    (let ((chars (string->list s))
+	  (string-var (gensym "string")))
```
Comment (wingo) [inline schism/compiler.ss]: Tabs here as well, fwiw.

File: schism/compiler.ss [hunk 7]
```diff
@@ -953,6 +1015,98 @@
...
+        (let ((tag (car expr)))
+          (cond
+           ((eq? tag 'null) expr)
+           ((eq? tag 'bool) expr)
+           ((eq? tag 'char) expr)
+           ((eq? tag 'begin)
+	    (cons 'begin (map lower-strings-expr (cdr expr))))
+           ((eq? tag 'var) expr)
+           ((eq? tag 'number) expr)
+           ((eq? tag 'eof-object) expr)
```
Comment (wingo) [inline schism/compiler.ss]: I think this isn't a possible tag -- eof-object is an icall.

File: schism/compiler.ss [hunk 8]
```diff
@@ -953,6 +1015,98 @@
...
+	   ((eq? tag 'eq?) `(eq?
+			     ,(lower-strings-expr (cadr expr))
+			     ,(lower-strings-expr (caddr expr))))
+	   ((eq? tag '<) `(<
+			   ,(lower-strings-expr (cadr expr))
+			   ,(lower-strings-expr (caddr expr))))
+	   ((eq? tag 'neq?) `(neq?
+			      ,(lower-strings-expr (cadr expr))
+			      ,(lower-strings-expr (caddr expr))))
+	   ((eq? tag 'bool) expr)
```
Comment (wingo) [inline schism/compiler.ss]: Bool already handled.  zero, eq?, <, and neq? are icalls.

File: schism/compiler.ss [hunk 9]
```diff
@@ -953,6 +1015,98 @@
...
+	   (let ((name (car b))
+		 (expr (cadr b)))
+	     `(,name ,(lower-strings-expr expr))))
+	 bindings))
+  (define (lower-string s)
+    (let ((chars (string->list s))
+	  (string-var (gensym "string")))
+      `(let ((,string-var (call make-string (number ,(length chars)))))
+	 (begin
+           (begin . ,(for-index (lambda (i c)
```
Comment (wingo) [inline schism/compiler.ss]: Might as well avoid the list and make for-index be vector-for-each or something.

**Comment by eholk** [comment]
This seems obsolete now in light of the anyref stuff.

**Comment by wingo** [comment]
> This seems obsolete now in light of the anyref stuff.

Indeed.  My apologies for encouraging exploration of this rabbit-hole!

**Comment by eholk** [comment]
No worries! What's code there for if not to delete it?

**Pull request closed without merging**